Improve pytest plugin documentation

gopiotr opened the pull request:
Improve documentation about pytest integration with Twister. Add examples of usage, improve description of available options and introduce automatic doc generation of two plugin classes (DeviceAdapter and Shell) basing on their docstrings from source code.

Generated documentation can be found here:
https://builds.zephyrproject.io/zephyr/pr/65631/docs/develop/test/pytest.html

gopiotr committed:
doc: test: improve pytest documentation

Improve documentation about pytest integration with Twister. Add
examples of usage, improve description of available options and
introduce automatic doc generation of two plugin classes (DeviceAdapter
and Shell) basing on their docstrings from source code.

Signed-off-by: Piotr Golyzniak <piotr.golyzniak@nordicsemi.no>

gopiotr commented:
@kartben - I would like to add autogenerating of documentation for two `pytest-twister-harness` classes (`DeviceAdapter` and `Shell`). To make it possible I need to add `pytest` package to `doc/requirements.txt` file - otherwise CI responsible for Documentation Build rised and error: https://github.com/zephyrproject-rtos/zephyr/actions/runs/6961647332/job/18943699575. 

Is this acceptable and good idea?

gopiotr commented:
FYI - generated documentation can be found here:
https://builds.zephyrproject.io/zephyr/pr/65631/docs/develop/test/pytest.html

gopiotr commented:
With rebase I added FAQ section.

gopiotr commented:
Rebased to fix typo in link.

kartben commented:
> @kartben - I would like to add autogenerating of documentation for two `pytest-twister-harness` classes (`DeviceAdapter` and `Shell`). To make it possible I need to add `pytest` package to `doc/requirements.txt` file - otherwise CI responsible for Documentation Build rised and error: https://github.com/zephyrproject-rtos/zephyr/actions/runs/6961647332/job/18943699575.
> 
> Is this acceptable and good idea?

No objection at all, but there might be a better place to add this dependency. For people building the docs locally, I think that as per our documentation we expect that they have a working Zephyr dev environment already. The reason that your PR wouldn't build in CI is that the GH action is trying to be smart and fast by only installing west, pretty much, but maybe we should have it also install scripts/requirements.txt? Another option would be to install just pytest in the GH action (https://github.com/zephyrproject-rtos/zephyr/blob/69d62add980bcafa6d90b73920372390cc10ea7c/.github/workflows/doc-build.yml#L191-L192)
 
@gmarull do you have a preference/suggestion here?

PerMac reviewed (changes requested):
only a minor request

gopiotr commented on doc/develop/test/pytest.rst:
```diff
@@ -91,13 +125,14 @@ keyword placed under ``harness_config`` section.
+Provide an `Shell <shell_class_>`_ class object with methods used to interact
+with shell application. It calls ``wait_for_promt`` method, to not start
+scenario until DUT is ready. Note that it uses ``dut`` fixture, so ``dut`` can
+be skipped when ``shell`` is used. Scope of this fixture is determined by the
```
Done. I changed "The shell fixture inherits from..." to "The shell fixture calls...", because I think that this better describes what is done from technical point of view.

gopiotr commented:
Rebased to solve merge conflict.

kartben commented on doc/develop/test/pytest.rst:
```diff
@@ -46,43 +46,75 @@ sets the test result accordingly.
+
+An example of a pytest test is given at
+:zephyr_file:`samples/subsys/testsuite/pytest/shell/pytest/test_shell.py`.
+Basing on configuration from the ``testcase.yaml`` file, Twister builds
+application from ``src`` and then if .yaml file contains ``harness: pytest``
+entry, Twister calls pytest in separate subprocess. Exemplary configuration
+.yaml file can looks like:
```
```suggestion
Using the configuration provided in the ``testcase.yaml`` file, Twister builds
the application from ``src`` and then, if the .yaml file contains a ``harness: pytest`` entry,
it calls pytest in a separate subprocess. A sample configuration file may look like this:
```

kartben commented on doc/develop/test/pytest.rst:
```diff
@@ -46,43 +46,75 @@ sets the test result accordingly.
+         harness: pytest
+         tags: foo
+
+By default, called pytest tries to look for tests in ``pytest`` directory,
```
```suggestion
By default, pytest tries to look for tests in a ``pytest`` directory
```

kartben commented on doc/develop/test/pytest.rst:
```diff
@@ -137,6 +171,135 @@ For example, here is a test with a fixture ``mcumgr``
+FAQ
+***
+
+1. **How to flash/run application only once per pytest session?**
```
I don't think numbering is necessary + it's prob better to actually keep the hierarchy using a proper heading
```suggestion
How to flash/run application only once per pytest session?
==========================================================
```

kartben commented on doc/develop/test/pytest.rst:
```diff
@@ -137,6 +171,135 @@ For example, here is a test with a fixture ``mcumgr``
+      harness_config:
+         pytest_dut_scope: session
+
+   more info can be find :ref:`here <pytest_dut_scope>`.
```
```suggestion
   More info can be found :ref:`here <pytest_dut_scope>`.
```

kartben commented on doc/develop/test/twister.rst:
```diff
@@ -505,14 +505,36 @@ harness_config: <harness configuration options>
-        when test case begin to running, default pytest directory is pytest.
-        After pytest finished, twister will check if this case pass or fail according
-        to the pytest report.
+        Specify a list of pytest directories, files or subtests that need to be
+        executed when test case begin to running, default pytest directory is
+        pytest. After pytest finished, twister will check if this case pass or
+        fail according to the pytest report. List of exemplary and valid
+        pytest roots is presented below:
```
```suggestion
        Specify a list of pytest directories, files or subtests that need to be
        executed when a test case begins to run. The default pytest directory is
        ```pytest``. After the pytest run is finished, Twister will check if 
        the tst case passed or failed according to the pytest report.
        As an example, a list of valid pytest roots is presented below:
```

kartben commented on doc/develop/test/pytest.rst:
you may use up to 100 characters per line. This is the case towards the beginning of this document but it looks like your changes are 80.

gopiotr replied:
Okay, I didn't know that the limit is 100 characters. So, I changed it back to 100 characters.

kartben commented on doc/develop/test/pytest.rst:
```diff
@@ -93,13 +125,15 @@ keyword placed under ``harness_config`` section.
-Note that it uses ``dut`` fixture, so ``dut`` can be skipped when ``shell`` is used.
-Scope of this fixture is determined by the ``pytest_dut_scope``
-keyword placed under ``harness_config`` section.
+Provide an `Shell <shell_class_>`_ class object with methods used to interact
```
```suggestion
Provide a `Shell <shell_class_>`_ class object with methods used to interact
```

kartben reviewed (changes requested):
I might need another pass but here's some feedback already. Thanks for this!

gopiotr committed:
samples: pytest: add README

Add README.rst file with description of sample which shows how to use
pytest and Twister together.

Signed-off-by: Piotr Golyzniak <piotr.golyzniak@nordicsemi.no>

gopiotr commented on doc/develop/test/pytest.rst:
```diff
@@ -135,6 +170,50 @@ For example, here is a test with a fixture ``mcumgr``
+Exemplary pytest tests in Zephyr project
+****************************************
+
+1. Shell demo - :zephyr_file:`samples/subsys/testsuite/pytest/shell`
```
Added

gopiotr commented on doc/develop/test/pytest.rst:
```diff
@@ -46,41 +46,75 @@ sets the test result accordingly.
-   --pytest-args='-k test_shell_print_version'
+There are two ways for passing extra arguments to the called pytest subprocess:
 
+1. From .yaml file, using ``pytest_args`` placed under ``harness_config``
```
Changed

nashif commented:
@kartben  please revisit

kartben commented on doc/develop/test/pytest.rst:
```diff
@@ -137,6 +166,132 @@ For example, here is a test with a fixture ``mcumgr``
+How to run only one particular test from a python file?
+=======================================================
+
+   This can be achieved in several ways. In .yaml file it can be added ``pytest_root`` keyword
```
```suggestion
   This can be achieved in several ways. In .yaml file it can be added using a ``pytest_root`` entry
```

kartben commented on samples/subsys/testsuite/pytest/shell/README.rst:
```diff
@@ -0,0 +1,79 @@
+.. zephyr:code-sample:: pytest_shell
+   :name: Pytest shell application test
+
+   Run shell application and execute pytest tests.
```
```suggestion
   Execute pytest tests against the Zephyr shell.
```

kartben commented on scripts/pylib/pytest-twister-harness/src/twister_harness/device/device_adapter.py:
```diff
@@ -25,7 +25,11 @@
 class DeviceAdapter(abc.ABC):
-    """Class defines an interface for all devices."""
+    """
+    Class defines an common interface for all device types (hardware,
```
```suggestion
    This class defines a common interface for all device types (hardware,
```

kartben commented on doc/develop/test/twister.rst:
```diff
@@ -505,14 +505,36 @@ harness_config: <harness configuration options>
+
+        .. code-block:: yaml
+
+          harness_config:
+            pytest_root:
+              - "pytest/test_shell_help.py"
+              - "../shell/pytest/test_shell.py"
+              - "/tmp/test_shell.py"
+              - "~/tmp/test_shell.py"
+              - "$ZEPHYR_BASE/samples/subsys/testsuite/pytest/shell/pytest/test_shell.py"
+              - "pytest/test_shell_help.py::test_shell2_sample"  # select pytest subtest
+              - "pytest/test_shell_help.py::test_shell2_sample[param_a]"  # select pytest parametrized subtest
```
looks like this might not be correctly indented (needs an extra 1-character positive indend), not sure if this causes rendering issues (the CI build expired) but prob worth fixing anyway

gopiotr replied:
I added extra indend - it is rendered properly (now and before too)

kartben commented on samples/subsys/testsuite/pytest/shell/README.rst:
thanks for adding the README! note for a future update: would be great if the `basic` sample also had one :)

gopiotr replied:
Yes, I agree that this could be valuable. At this moment I don't have enough time to do this even in this PR. Maybe in the future in separate one.

kartben commented on samples/subsys/testsuite/pytest/shell/README.rst:
```diff
@@ -0,0 +1,79 @@
+.. zephyr:code-sample:: pytest_shell
+   :name: Pytest shell application test
```
```suggestion
   :name: Pytest shell application testing
```

kartben commented on doc/develop/test/twister.rst:
```diff
@@ -505,14 +505,36 @@ harness_config: <harness configuration options>
+        Specify a list of pytest directories, files or subtests that need to be
+        executed when a test case begins to run. The default pytest directory is
+        ``pytest``. After the pytest run is finished, Twister will check if
+        the tst case passed or failed according to the pytest report.
```
```suggestion
        the test case passed or failed according to the pytest report.
```

kartben reviewed (changes requested):
sorry that it took me so long to have another pass.

gopiotr commented:
Thank you for review @kartben :) Please re-review

gopiotr commented:
Rebased to main
